dispatcher/backend/src/migrate-data-mongo-to-pg.py

```diff
@@ -0,0 +1,287 @@
+
+
+def pre_checks_ok_internal(session: so.Session) -> bool:
+    pre_checks_ok = True
+
+    count = session.query(dbm.User).count()
+    if count > 0:
+        logger.warning(
+            f"Please empty user table before migration ({count} records found)"
+        )
+        pre_checks_ok = False
+
+    count = session.query(dbm.Sshkey).count()
+    if count > 0:
+        logger.warning(
+            f"Please empty sshkey table before migration ({count} records found)"
+        )
+        pre_checks_ok = False
+
+    count = session.query(dbm.Refreshtoken).count()
+    if count > 0:
+        logger.warning(
+            "Please empty refresh tokens table before migration "
+            f"({count} records found)"
+        )
+        pre_checks_ok = False
+
+    return pre_checks_ok
+
+
+class Migrator(ABC):
+    _commit_every: int
+
+    def __init__(self, commit_every: int = 1000) -> None:
+        self._commit_every = commit_every
+
+    @abstractmethod
+    def get_object_name(self) -> str:
+        pass
+
+    @abstractmethod
+    def get_mongo_collection(self) -> BaseCollection:
+        pass
+
+    @abstractmethod
+    def get_sql_tables(self) -> List[Tuple[str, dbm.Base]]:
+        pass
+
+    @abstractmethod
+    def get_expected_mongo_keys(self) -> List[str]:
+        pass
+
+    @abstractmethod
+    def get_expected_mongo_json_objects(self) -> List[str]:
+        pass
+
+    @abstractmethod
+    def migrate_one_doc(self, session: so.Session, mongo_obj) -> None:
+        pass
+
+    def get_next_mongo_doc(self):  # -> Generator[Any | None, None, None]:
+        for mongoObj in self.get_mongo_collection().find({}, batch_size=100):
+            yield mongoObj
+        yield None
+
+    def _key_starts_with(self, key, items):
+        for item in items:
+            if key.startswith(item):
+                return True
+        return False
+
+    def check_one_doc(self, mongo_obj):
+        existing_keys = KeysExporter.get_keys(mongo_obj)
+        expected_keys = self.get_expected_mongo_keys()
+        expected_json_objects = self.get_expected_mongo_json_objects()
+        for existing_key in existing_keys:
+            if existing_key not in expected_keys and not self._key_starts_with(
+                existing_key, expected_json_objects
+            ):
+                raise Exception(
+                    f"Unexpected key {existing_key} found in "
+                    f"{self.get_object_name()} object id {str(mongo_obj['_id'])}"
+                )
+
+    def migrate(self) -> None:
+        logger.info(
+            f"\n##################\nMigrating {self.get_object_name()}"
+            "\n##################"
+        )
+        count = self.get_mongo_collection().count_documents({})
+        logger.info(f"{count} {self.get_object_name()} found in MongoDB")
+
+        cnt = 1
+        next_mongo_doc = self.get_next_mongo_doc()
+
+        with Session() as session:
+            while True:
+                with session.begin():
```

Comment:
Wouldn't a `begin_nested()` be more appropriate? Just asking

Reply:
Nope, `begin_nested()` is issuing a savepoint, i.e. transactions before the statement are not committed and objects are still managed by the ORM. The savepoint / begin_nested is useful only when you want to be able to rollback changes only up to the savepoint instead of the whole transaction.

Here I want to commit into the database chuncks of data in order to avoid having too many objects managed by the ORM and hence stored in memory. Meaning that they will be commited to the database. It has the drawback that should we have any error at some point later in the processing, some data will already be stored in the database. But this is not an expected thing since the migration will be tested many times. Moreover, should we have any transient error (e.g. network issue between the migration tool and the DB) which will stop the migration in the middle of the process, I would prefer to prune the DB and restart the migration from scratch instead of relying on retry logics.